### PR TITLE
check sagemaker training job status before deferring SageMakerTrainingOperator

### PR DESCRIPTION
**Pull request by Lee-W**
While running a sagemaker training job in deferrable mode, the condition might already be met before we defer the task into the trigger. This PR intends to check the job status before deferring the task to trigger.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/sagemaker.py [hunk 1]
```diff
@@ -1300,3 +1300,13 @@ def create_auto_ml_job(
             if "BestCandidate" in res:
                 return res["BestCandidate"]
         return None
+
+    async def describe_training_job_async(self, job_name: str) -> dict[str, Any]:
```
Comment (pankajkoti) [inline airflow/providers/amazon/aws/hooks/sagemaker.py:1316]: Where does this get used?

File: airflow/providers/amazon/aws/operators/sagemaker.py [hunk 2]
```diff
@@ -1085,8 +1089,50 @@ def execute(self, context: Context) -> dict:
...
+            if self.print_log:
+                instance_count = description["ResourceConfig"]["InstanceCount"]
+                last_describe_job_call = time.monotonic()
+                job_already_completed = status not in self.hook.non_terminal_states
+                _, last_description, last_describe_job_call = self.hook.describe_training_job_with_log(
+                    self.config["TrainingJobName"],
+                    {},
+                    [],
+                    instance_count,
+                    LogState.TAILING if job_already_completed else LogState.COMPLETE,
```
Comment (pankajkoti) [inline airflow/providers/amazon/aws/operators/sagemaker.py]: Can we recheck if the conditional is correct here?

File: airflow/providers/amazon/aws/hooks/sagemaker.py (hunk 1 above)
Comment (Lee-W) [reply airflow/providers/amazon/aws/hooks/sagemaker.py:1316]: I refined the code a bit and used it in the trigger. Could you please take a look? Thanks!

File: airflow/providers/amazon/aws/operators/sagemaker.py (hunk 2 above)
Comment (Lee-W) [reply airflow/providers/amazon/aws/operators/sagemaker.py]: Yes, i think the condition is correct. Is there any concern here?

Comment (pankajkoti) [reply airflow/providers/amazon/aws/operators/sagemaker.py]: I was thinking if the condition should be other way round. Tail the logs while the job is not complete and if complete then set LogState.Complete.

Comment (Lee-W) [reply airflow/providers/amazon/aws/operators/sagemaker.py]: yep, thanks for the reminder. fixed it

**Review by pankajkoti — APPROVED** [review]
Looks like a static check is failing

**Review by phanikumv — APPROVED** [review]

**Pull request merged**